lib/recursive-methods.rb

```diff
@@ -1,49 +1,96 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of n.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 1 || n == 0
+  
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of the string.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  
+  reversed_s = reverse(s[1..-1])
+  reversed_s += s[0]
+  return reversed_s
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n), the amount of recursive calls depends on the size of the string.
+# Space complexity: O(n), each recursive call is added to the stack. 
+def reverse_inplace(s, low = 0, high = s.length-1)
+  return s if high-low <= 1
+  
+  temp = s[low]
+  s[low] = s[high]
+  s[high] = temp
+  
+  return reverse_inplace(s, low+1, high-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of n.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 if n == 1
+  
+  return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of the string.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length == 0
+  
+  if s[0] == "(" && s[-1] == ")"
+    return nested(s[1..s.length-2])
+  else 
+    return false
+  end
+  
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(log n), the amount of searches is cut in half with every recursive call.
+# Space complexity: O(n), each recursive call is added to the stack. 
+def search(array, value, low = 0, high = array.length-1)
```

Comment:
👍 The time/space complexity is O(log n) in both.

This also assumes that the array is sorted.